### PR TITLE
JV880 support

### PR DESCRIPTION
**Pull request by giulioz**
Currently only shows the intro animation

### CODE REVIEW

**Comment by Grieferus** [comment]
It's not SC-55-related synth, so cast it out for now.

**Comment by giulioz** [comment]
What do you mean? It has exactly the same architecture. It's possible to support it with minimal changes, but if it's off-topic for this project I think it's up for the original author to decide.

**Comment by zaphod77** [comment]
jv-880 i believe has hardware similar to sc-880, or is it the other way around?

so i wouldn't expect to get much further.

**Comment by giulioz** [comment]
@zaphod77 it's not, it has the exact same PCM chip
You can verify it here: https://www.synthxl.com/wp-content/uploads/2017/12/roland-jv-880-Service-Notes.pdf
Compare the difference with the SC55: https://data2.manualslib.com/pdf5/102/10146/1014558-roland/sound_canvas_sc55mkii.pdf?f5d24e3fcee0bc78a3fc0292d6b862da

**Comment by Grieferus** [comment]
Still, while the CPU is the same, the waveform ROMs are pretty much obscure for us. Besides, this emulator is for SC-55-based synthesizers, not JV (Unless it's JV-30, which is basically SC-55 with keyboard) or laters Sound Canvas modules.

**Comment by giulioz** [comment]
I've just dumped the waveform ROMs and they use exactly the same format.
The JV880 is an SC-55 based synthesizer.

**Comment by pachuco** [comment]
I don't know. The familly is expanding in cool and unexpected directions.

**Comment by nukeykt** [comment]
JV 880 use same CPU and PCM chip, so it fits the project scope IMO

**Comment by Grieferus** [comment]
Okay, our Lord and Saviour has spoken, so I must comply!

**Comment by zaphod77** [comment]
wow. roland does love to make things confusing, don't they.  

so all the magic really is in that pcm chip?

guess that's why it 's so affordable.

jv-30 as well seems to use the same cpu (H8/532), and its' a sc-55 with a keyboard essentially.

**Comment by nukeykt** [comment]
MCU firmware is what makes SC-55, well, a SC-55.
MCU firmware does all midi processing, PCM chip voices management, pitch calculaions/LFOs, envelope/filter control, UI/LCD/buttons and whatnot.
PCM chip on other hand is rompler with fixed-function effects (sample interpolation, amplitude, filter, chorus/reverb) and envelope interpolators.

**Comment by michael02022** [comment]
I have to remind the JV-880 supports Expansion Boards, is it possible to load them with this variant? That would be rad ngl

**Comment by giulioz** [comment]
To support expansion boards it should be sufficient to just append them at the end of the waveform rom, so definitely possible

**Comment by Axis4s** [comment]
🎊 
wishing both this new 880 and sc55 emu get VSTi versions eventually so they can more easily be integrated with audio software that has VSTi support

**Comment by Axis4s** [comment]
also whats the original filenames for the 880 firmware? before i have to rename them to work with the 880 emu mode

**Comment by giulioz** [comment]
Expansion cards are now working as well: https://www.youtube.com/watch?v=d97a_QnxYsg

File: src/mcu_opcodes.cpp [hunk 1]
```diff
@@ -330,6 +330,23 @@ void MCU_Jump_RTS(uint8_t operand)
...
+    int16_t imm = (int8_t)MCU_ReadCodeAdvance();
+    mcu.pc = MCU_PopStack();
+
+    if (operand == 0x14)
+    {
+        mcu.r[7] += imm;
+        if (mcu.r[7] & 1)
+            MCU_ErrorTrap();
+    }
+    else if (operand == 0x1c)
```
Comment (nukeykt) [inline src/mcu_opcodes.cpp:344]: `if` should be removed to catch any other potential cases

File: src/mcu_opcodes.cpp [hunk 2]
```diff
@@ -342,6 +359,13 @@ void MCU_Jump_JMP(uint8_t operand)
             mcu.cp = (uint8_t)MCU_PopStack();
             mcu.pc = MCU_PopStack();
         }
+        else if (opcode_h == 0x19)
+        {
+            MCU_PushStack(mcu.pc);
+            MCU_PushStack(mcu.cp);
+            mcu.cp = mcu.r[opcode_l] & 0xff;
```
Comment (nukeykt) [inline src/mcu_opcodes.cpp:366]: odd case of opcode_l should be handled somehow, I guess hardware would just ignore LSB, so `opcode_l &= ~1` will work.

File: src/mcu_opcodes.cpp [hunk 3]
```diff
@@ -1254,6 +1294,15 @@ void MCU_Opcode_SHLR(uint8_t opcode, uint8_t opcode_reg)
         MCU_SetStatus(C, STATUS_C);
         MCU_SetStatusCommon(data, operand_size);
     }
+    else if (opcode_reg == 0x01 && operand_type != GENERAL_IMMEDIATE) // SHAR
+    {
+        uint32_t data = MCU_Operand_Read();
+        uint32_t C = data & 0x1;
+        data >>= 1;
```
Comment (nukeykt) [inline src/mcu_opcodes.cpp:1301]: MSB (bit 7 in byte mode, bit 15 in word mode) should be preserved according to manual

**Review by nukeykt — CHANGES_REQUESTED** [review]

File: src/mcu_opcodes.cpp (hunk 3 above)
Comment (giulioz) [reply src/mcu_opcodes.cpp:1318]: good catch! fixed, I hope the fix is more correct

File: src/mcu_opcodes.cpp (hunk 1 above)
Comment (giulioz) [reply src/mcu_opcodes.cpp:344]: Yeah but 0x1c is the word case that is not implemented, I left it with a comment for the future and also added another else at the bottom if that's ok

File: src/mcu.cpp [hunk 4]
```diff
@@ -83,6 +84,12 @@ const char* roms[ROM_SET_COUNT][5] =
     "cm300_waverom1.bin",
     "cm300_waverom2.bin",
     "cm300_waverom3.bin",
+
+    "jv880_rom1.bin",
+    "jv880_rom2.bin",
+    "jv880_waverom1.bin",
+    "jv880_waverom2.bin",
+    "SR-JV80-01 Pop - CS 0x3F1CF705.bin",
```
[the remark this answers is not in the document]
Comment (giulioz) [reply src/mcu.cpp]: renamed, also made it optional so that it can run without it

**Comment by giulioz** [comment]
Also updated the readme adding the keymapping and rom info

**Review by nukeykt — APPROVED** [review]

**Comment by nukeykt** [comment]
thanks!

**Pull request merged**